lblod/scheduled-job-controller-service
# Authentication with scheduled jobs

# Description
DL-5039

This PR is about fixing the scheduled jobs so they can run authenticated jobs.

- Ensures the source collection (scheduled job) is properly cloned to the clone collection (normal job) so it can share the authentication configuration informations. **DONE**
- Deletes the source collection authentication configuration content when the scheduled job is deleted. **DONE**
- Adds encryption/decrytion for the authentication configuration. **DONE**


## Context 

There is cases where we need to harvest data from vendors, they're usually using an authentication mechanism. We need this feature to pass the credentials when a job is created from a scheduled job. 

What service does now  : 
- The scheduled job will serve as template to make a clone (normal job), meaning the source collection (from the scheduled job) containing authentication configuration relationship is attached to the normal job's collection.

# Type of change


# Related services

- [app-worship-harvester](https://github.com/lblod/app-worship-harvester)
- [scheduled-job-controller-service](https://github.com/lblod/scheduled-job-controller-service)
- [frontend-harvesting-self-service](https://github.com/lblod/frontend-harvesting-self-service)

# How to test

`app-worship-harvester` needs authentication.

Please make sure you have in your migration folder : 2023/local/`20230309104958-add-mock-user.sparql`.
You can find this migration on the development server from the `app-worship-harvester-qa`. Also you need to enable from the frontend container the variable `EMBER_AUTHENTICATION_ENABLED: 'true'`

You will need to create a docker-compose.override.yml file with the following image : `poltergeistz/scheduled-job-controller-service-local:latest`

Here's a base you can follow :

```yml
version: "3.7"

services:
  frontend:
      image: lblod/frontend-harvesting-self-service:2.0.0
      environment:
        EMBER_AUTHENTICATION_ENABLED: 'true'
  scheduled-job-controller-service:
      image: poltergeistz/scheduled-job-controller-service-local:latest
      environment:
        SECRET_KEY: ""
        IV: ""
``` 

For the `SECRET_KEY` and `IV` you can set any values you want but it has to be 32 (SECRET_KEY) and 16 (IV) character long.

For the login information it should be the same as QA.

You can now start the app and login.

## What to check

- When creating a new scheduled job, check from the source collection if the authentication configuration has encrypted data. You can use the yasgui with this query to check. You can copy the authentication configuration uri from the scheduled job you've created in ember inspector to then paste it into the query. :heavy_check_mark: 

BASIC_AUTH

```sparql
PREFIX dgftSec: <http://lblod.data.gift/vocabularies/security/>
PREFIX meb: <http://rdf.myexperiment.org/ontologies/base/>
PREFIX muAccount: <http://mu.semte.ch/vocabularies/account/>
SELECT DISTINCT * WHERE {
  <REPLACE_THIS_BY_THE_AUTHENTICATION_CONFIGURATION_URI> dgftSec:secrets ?secrets .
          ?secrets meb:username ?user ;
            muAccount:password ?pass .
} 
```

OAUTH2

```sparql
PREFIX dgftSec: <http://lblod.data.gift/vocabularies/security/>
PREFIX dgftOauth: <http://kanselarij.vo.data.gift/vocabularies/oauth-2.0-session/>
PREFIX wotSec: <https://www.w3.org/2019/wot/security#>
SELECT DISTINCT * WHERE {
   <REPLACE_THIS_BY_THE_AUTHENTICATION_CONFIGURATION_URI> dgftSec:secrets ?secrets .
          ?secrets dgftOauth:clientId ?clientId ;
            dgftOauth:clientSecret ?clientSecret .
   <REPLACE_THIS_BY_THE_AUTHENTICATION_CONFIGURATION_URI> dgftSec:securityConfiguration ?scheme .
         ?scheme wotSec:token ?token ;
            wotSec:flow ?flow .
}
```

- When the scheduled job is created check if the job is successfull for both oauth2 and basic auth. :heavy_check_mark: 
- When deleting a scheduled job the authentication configuation from the source collection should not contain any credentials data. :heavy_check_mark: 

# Links to other PR's

- https://github.com/lblod/frontend-harvesting-self-service/pull/27
